### PR TITLE
✨  Favicon URI

### LINKED ISSUE #7688: Feature: Blog Icon
In external sites, we want to be able to display an icon for each blog. This is effectively the same problem as in [issue #4869](https://github.com/TryGhost/Ghost/issues/4869) - we need a static URL at which an external service such as an RSS feed or ghost.org can access to fetch a representative URL.

This icon should be used as the favicon for the blog. It should also be used when any external service looks for an icon. Browsers always look for a `/favicon.ico` file, at the root of the current domain. This means the behaviour of this file will always be slightly different for blogs which use a subdirectory. When rendering HTML, browsers use 

Ideally, we want to serve a .png file. I'm not 100% on what we can or can't do here, and Googling didn't turn up much. Most browsers support a `/favicon.png` , and it'll be fine to use this for browsers & ghost.org, but to ensure we also capture the RSS issue, it may be worth 301-redirecting favicon.ico -> favicon.png for blogs which have an icon uploaded. This is an unknown, and will require a little bit of experimenting with. Any solution which works in feedburner should be considered acceptable for this task.

**Image Requirements** 

- Only permit PNG images (client + server?)
- Enforce that the image must be square (server side only?)
- Dimensions between 32px and 1000px (server side only?)
- Size not larger than 100kb (server side only?)

### Part A: Getting the icon uploaded:

The icon will be a 3rd image on the general settings screen in Ghost. The main difficulty here is implementing the image validation, to ensure the image meets our requirements. At least for now, this can most easily be done server side, I believe.

- Add a new general setting in Ghost for blog icon
- Add an additional image upload UI to the /settings screen
        - The icon should only allow uploads, not urls - so the url feature needs to be disabled
- Ensure that when uploaded the image is validated:
	- First check the filesize, and reject anything over 100kb, this means that further processing should at least be quick & easy!
	- Use our image-size utility to check that the size is square and within the right dimensions
	- If any check fails then fully detailed error messages should be returned explaining to the user exactly why an image is not suitable
- The client will need to display any validation errors returned from the server.
- This image will likely be small, we will need to ensure that the image upload modals in general/settings do show the trash icon!

### Part B: Using the icon  

Any blog that has an uploaded icon should use this instead of any theme favicons. 

 **Note** : the existing favicon implementation works as described in the [asset helper documentation](http://themes.ghost.org/docs/asset#section-favicons) . It is odd & fiddly and needs overhauling. The existing implementation requires users to output their own meta tag, as Ghost 1.0 will need to output the meta tag itself, this will constitute a breaking change to the theme API.

Ghost will need an internal concept of what its "current favicon" is. This is possibly the 2nd MOST! tricky part of this whole implementation. In the case that a favicon has been uploaded, it's super easy, but otherwise, it requires some work to detect favicons in the active theme (and this must be refreshed when a theme is activated or overridden). 

The ideal order of precedence for the favicon for a blog, is:

- uploaded favicon
- activetheme/assets/favicon.png
- activetheme/assets/favicon.ico
- activetheme/favicon.png
- activetheme/favicon.ico
- Ghost's favicon.png file (this is a new default, that needs to be added)
- Ghost's favicon.ico file (not sure if we need this? maybe for older browsers)

 **Outputting the favicon in themes:** 

The `{{ghost_head}}` helper should be updated to output a new meta tag. 

In the case that the "current favicon" is a .png file:

 `<link rel="shortcut icon" href="favicon.png" type="image/png" />` 

In the case that the "current favicon" is a .ico file:

 `<link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />` 

 **Note** : It would be good to experiment here with what will happen for themes that already have this output above `{{ghost_head}}` i.e. Casper.

 **Note: ** I don't know if the `{{amp_head}}` should output something different?

As well as outputting the meta tag with `{{ghost_head}}` , we should also make sure that the icon url (e.g. favicon.png) is available as `{{@blog.icon}}` .

 **Outputting the icon on a URI** 

In order for the icon to work for external sites, we need to render the icon whenever it is accessed via a URI, e.g. `/favicon.png` 

The existing favicon implementation uses a very odd piece of logic to serve any requests for `/favicon.ico` using our [shared file helper](https://github.com/TryGhost/Ghost/blob/4411f8254f90f47e4dff02963afb2131983a4462/core/server/blog/app.js#L44-L44) , which is totally wrong for this usecase - this code should basically be deleted and forgotten about, and a new piece of middleware for favicons should be written.

The new middleware should live in the exact same place in the stack (e.g. after the two themeHandler middlewares) and the logic should be as follows:

- The middleware should be aware of what the "current favicon" is. 
- If the "current favicon" is a .png file, ensure it is served from `/favicon.png` . Also, 301 redirect any request for `/favicon.ico` to `/favicon.png` (this may need testing with RSS readers to ensure it works).
- If the "current favicon" is a .ico file (i.e. No icon was uploaded), ensure it is served from `/favicon.ico` . In order for this case to work for external sites, we also need to either 301 redirect `/favicon.png` to `/favicon.ico` OR just serve the Ghost default icon.

In terms of implementation, this middleware should be small and super fast, it's going to need to read the contents of the icon file, and then cache those contents. 

**Update**: The existing favicon implementation assumes that it's ok to read from the file system, totally ignoring our storage adapter layer. The new solution needs a `read` method added to the storage adapters (see #7687) so that if a file has been uploaded, we are able to access it. If no file has been uploaded, then for now it is ok to read from the file system, as currently we only support local file system storage for themes & the shared files.

We then need to manage ensuring that the cache is cleared at the right times, so that if a user uploads a new logo, or deletes a logo, we correctly display the right one without requiring a restart of Ghost. Handling the caching is the 1st MOST! tricky part of this.

The rules are, to the best of my knowledge, as follows:

- on first load, we need to figure out the "current favicon" and store a reference to it. That reference will need to include:
	- the path to the favicon file = where the file itself can be opened and read
	- a reference to the type, either upload, theme or default (default is when we serve Ghost's favicon)
- For uploaded icons - the cache needs to be cleared when a new icon is uploaded, or the icon is deleted
- For theme icons - the cache needs to be cleared whenever the assetHash is cleared
- For default icons - the cache never needs to be cleared

## Useful links

Some places to use for inspiration for how to write the file reading and caching logic are:

- Ghost's [shared file middleware](https://github.com/TryGhost/Ghost/blob/0ae0a0b4900413a5c9bc6b694df6889bbcf7c701/core/server/middleware/serve-shared-file.js#L9-L9](https://github.com/TryGhost/Ghost/blob/0ae0a0b4900413a5c9bc6b694df6889bbcf7c701/core/server/middleware/serve-shared-file.js#L9-L9)) 
- The main [express serve favicon](https://github.com/expressjs/serve-favicon/blob/master/index.js) middleware (note this only supports favicon.ico)
- Ghost's logic for [refreshing the theme cache](https://github.com/TryGhost/Ghost/blob/6473c9e858342dba214ac3483a5918afee7e486b/core/server/api/index.js#L79-L79) 
- Ghost's weird [theme settings cache](https://github.com/TryGhost/Ghost/blob/d81bc91bd251bc4dc8d9c6aedb749f76c987ae18/core/server/api/settings.js#L62-L62) 
- Ghost's [theme handling middleware](https://github.com/TryGhost/Ghost/blob/59e2694acffb5a514e5857a8737b893f44284677/core/server/middleware/theme-handler.js#L49-L49) which deals with the assetHash and express's undocumented view cache

### PR DESCRIPTION
**Pull request by aileen**
closes #7688

Use `/favicon.ico` and `/favicon.png` in blog app. Depending on type of storage (custom upload = local file storage), serves either from storage adapter with `read()` method or reads the bytes via `fs`.

Redirects requests for `favicon.ico` to `favicon.png` if custom `png` icon is uploaded and vice versa.

Redirect requests for `favicon.png` to `favicon.ico` if default icon is used (in `core/shared`).

Changes the `{{asset}}` helper for favicon to not serve from theme assets anymore. It will either be served the custom blog-icon or the default one.

The `{{@blog.icon}}` helper renders the url of the **uploaded** blog icon. It won't render the default icon.

TODOs for #7688:
- [X] Figure out, which favicon should be used (uploaded or default) -> #7713
- [X] Serve and redirect the favicon for any browser requests, incl. redirects, output new favicon in `{{ghost_head}}` -> this PR
- [X] Upload favicon via `general/settings` and implement basic admin validations -> TryGhost/Ghost-Admin#397
- [X] Build server side validations -> #7893

### CODE REVIEW

File: core/server/helpers/ghost_head.js [hunk 1]
```diff
@@ -75,7 +76,7 @@ function ghost_head(options) {
     if (this.statusCode >= 400) {
         return;
     }
-
+    console.log(!config.get('theme:icon'));
```
Comment (kevinansfield) [inline core/server/helpers/ghost_head.js]: :trollface:

Comment (AileenCGN) [inline core/server/helpers/ghost_head.js]: Haha... I have no words ...

**Comment by kevinansfield** [comment]
@AileenCGN I'm seeing a duplicated icon link when testing this - is that expected? I can't seem to get Chrome to pick up the PNG version even after clearing my cache but Safari does show it.

screen shot 2017-01-26 at 17 18 40

**Comment by AileenCGN** [comment]
Yes, that's still a 'we-figure-out-what-will-happen' thing. The first link comes from the theme (which will be removed shortly) and the second link is now produces in `{{ghost_head}}`.

See in the spec:
>Note : It would be good to experiment here with what will happen for themes that already have this output above {{ghost_head}} i.e. Casper.

**Comment by AileenCGN** [comment]
Regarding the `.png` not working, do you mean a custom uploaded icon? That works fine for me in Chrome... 🤔  even without clearing the cache...

**Comment by kevinansfield** [comment]
> Regarding the .png not working, do you mean a custom uploaded icon?

Yes, it still always shows me the default icon however if I try visiting `/favicon.ico` directly it does redirect me to the png, might be worth a bit of investigation if it's repeatable for others - it will likely be a source of confusion for users uploading a new icon and not seeing it take affect.

**Comment by AileenCGN** [comment]
If you uploaded a `.png` file and you're request `favicon.ico` you should get redirected to `favicon.png` and vice versa. But yeah, it's weird that this happens to you. @kirrg001 didn't have this issue either... We both tested it manually.

**Pull request merged**